<launch>
  <!-- Arguments -->
  <arg name="model" default="waffle_pi"/>
  <arg name="bag_name" default="fixed_slam_easy"/>
  <arg name="map_file" default="$(find turtlebot3_datasets)/data/map.yaml"/>

  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- Map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

  <!-- TF -->
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf2_mocap2odom" required="true" args="0.935 1.34 -0.023 0.001 -0.003 0.737 0.676 mocap odom" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf2_mocap2map" required="true" args="0.935 1.34 -0.023 0.001 -0.003 0.737 0.676 mocap map" />
  
  
  <!-- ROSBAG -->
  <param name="/use_sim_time" value="true"/>
  <node pkg="rosbag" type="play" name="rosbag_player" 
          args="--clock $(find turtlebot3_datasets)/data/$(arg bag_name).bag"/>

  <!-- EKF robot localisation -->
  <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_odom" clear_params="true">
    <param name="frequency" value="10." />
    <param name="sensor_timeout" value="0.2" />
    <param name="two_d_mode" value="true" />
    <param name="publish_tf" value="false" />
    <param name="map_frame" value="map" />
    <param name="odom_frame" value="odom" />
    <param name="base_link_frame" value="base_link" />
    <param name="world_frame" value="odom" />
    <param name="odom0" value="odom" />
    <param name="scan0" value="scan" />
    
    <rosparam param="odom0_config">
      [true, true, false,
       false, false, true,
       true, true, false,
       false, false, true,
       false, false, false]</rosparam>
       
    <param name="odom0_relative" value="true"/>
    <param name="scan0_relative" value="true"/>
    <param name="imu0" value="imu" />
    
    <rosparam param="imu0_config">
      [false, false, false,
       false, false, true,
       false, false, false,
       false, false, true,
       false, false, false]</rosparam>
    
    <rosparam param="scan0_config">
      [true, true, false,
       false, false, true,
       false, false, false,
       false, false, false,
       false, false, false]</rosparam>
    <param name="imu0_relative" value="true"/>

    
    <param name="imu0_relative" value="true"/>
    
    <rosparam param="initial_state">
      [-2.0,-0.5,  0.0,
       0.0,  0.0,  0.0,
       0.0,  0.0,  0.0,
       0.0,  0.0,  0.0,
       0.0,  0.0,  0.0]</rosparam>
  </node>

  <!-- AMCL-->
  
  <!-- info for path -->
  <!-- params from http://wiki.ros.org/laser_scan_matcher?distro=fuerte -->
  <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" 
    name="laser_scan_matcher_node" output="screen">

    <!-- Topics -->
    <remap from="scan" to="scan"/>
    <remap from="imu/data" to="imu/data_compass"/>
    <remap from="odom" to="odom"/>

    <!-- Coordinate frames -->
    <param name="fixed_frame" value="odom" />
    <param name="base_frame" value="base_link" />

    <!-- Motion prediction -->
    <param name="use_imu" value="true"/>
    <param name="use_odom" value="true"/>

    <!-- Point cloud input -->
    <param name="use_cloud_input" value="false"/>
    <param name="cloud_range_min" value="0.4"/>
    <param name="cloud_range_max" value="10.0"/>

    <!-- Key frames -->
    <param name="kf_dist_linear" value="0.1"/>
    <param name="kf_dist_angular" value="0.175"/>

    <!-- Output -->
    <param name="publish_tf" value="true"/>
    <param name="publish_pose" value="false"/>
    <param name="publish_pose_stamped" value="true"/>

    <!-- Scan matching -->
    <param name="max_iterations" value="50"/>
    <param name="max_correspondence_dist" value="0.3"/>
    <param name="max_angular_correction_deg" value="45.0"/>
    <param name="max_linear_correction" value="0.5"/>
    <param name="epsilon_xy" value="0.000001"/>
    <param name="epsilon_theta" value="0.000001"/>
    <param name="outliers_maxPerc" value="0.90"/>
  </node>
  
  <!-- Groundtruth path -->
  <node pkg="turtlebot_sim" type="odom_to_path.py" name="odom_to_path" required="true" />
  
  <!-- Estimated path -->
  <node pkg="turtlebot_sim" type="laserscan_ekf_to_path.py" name="laserscan_ekf_to_path" required="true" />
  
  <!-- Position error publisher -->
  
  
  <!-- rviz -->
  <!--<node pkg="rviz" type="rviz" name="rviz" required="true"/>-->
  
  <node pkg="rviz" type="rviz" name="rviz" required="true" args="-d $(find turtlebot_sim)/rviz/turtle_sim_ekf_laser.rviz" />
    <node name="tf_to_path" pkg="tf_to_trajectory" type="tf_to_path.py">
    <param name="parent_frame_id" value="mocap"/>
    <param name="child_frame_id" value="mocap_laser_link"/>
    <param name="out_topic_name" value="/ground_truth"/>
    <param name="append_rate" value="5"/>
  </node>
  

</launch>
